docs/upgrading_to_nameko_3.rst

```diff
@@ -0,0 +1,347 @@
+.. _upgrading-to-nameko-3:
+
+Upgrading to Nameko 3
+=====================
+
+What's new in Nameko 3
+----------------------
+
+Configuration improvements
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+* Configuration loaded from a file or declared in the CLI is now available in a :ref:`global config object <nameko-config>`.
+* Configuration may now be :ref:`declared on the CLI <cli-define-option>`.
+* Configuration access via the service container is now deprecated.
+* Builtin extensions and other facilities that read the configuration now access it via the global.
+
+CLI improvements
+~~~~~~~~~~~~~~~~
+
+* The CLI has been rebuilt using `click <http://click.pocoo.org/>`_ to be more robust, reliable and extensible..
+* The `backdoor` command has been removed; `run`` informs about how to connect to the backdoor port.
+
+AMQP refactors
+~~~~~~~~~~~~~~
+
+The “built-in” extensions that use AMQP (RPC, Events) have been refactored. The API is backwards compatible but enhanced in many ways:
+
+* `prefetch_count` is exposed as a first-class parameter, rather than inheriting max_workers.
+* Publisher options can be provided to the EventDispatcher and Publisher.
+* ...
+
+Configuration improvements
+--------------------------
+
+.. _nameko-config:
+
+Global config object
+~~~~~~~~~~~~~~~~~~~~
+
+The configuration is now available in a global object, rather than being passed around in the service container. This means that configuration can be accessed from anywhere, including from outside the service container.
+
+The new `nameko.config` dictionary allows you to access service configuration at any time, including service class definition:
+
+.. code-block:: python
+
+    from nameko.messaging import Consumer
+    from nameko import config
+
+    class Service:
+
+        @consume(
+            queue=Queue(
+                exchange=config["MY_EXCHANGE"],
+                routing_key=config["MY_ROUTING_KEY"],
+                name=config["MY_QUEUE_NAME"]
+            ),
+            prefetch_count=config["MY_CONSUMER_PREFETCH_COUNT"]
+        )
+        def consume(self, payload):
+            pass
+
+As you can see having config available outside service container also helps with configuring entrypoints and dependency providers by passing options to their constructors:
+
+.. code-block:: python
+
+    from nameko import config
+    from nameko_sqlalchemy import Database
+
+    class Service:
+        db = Database(engine_options=config["DB_ENGINE_OPTIONS"])
+
+.. _cli-define-option:
+
+Declaration of config in the CLI
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Configuration can now be declared on the CLI, rather than only being loaded from a file. This is useful for setting configuration that is specific to the environment in which the service is running, such as the hostname of a database server.
```

Comment:
Please specify if the CLI definitions are config overrides. i.e. if a config file is provided, and still some options are defined on the CLI, which will take precedence?